sms/src/main/java/com/gruppo4/sms/network/SMSAbstractNetworkManager.java

```diff
@@ -0,0 +1,212 @@
+    static final String JOIN_AGREED = "JA";
+    static final String JOIN_PROPOSAL = "JP";
+    static final String SPLIT_CHAR = "_";
+
+    private SMSNetworkDictionary dict = new SMSNetworkDictionary<>();
+
+    // joinSent keeps track of JOIN_PROPOSAL requests still pending.
+    private ArrayList<SMSPeer> joinSent = new ArrayList<>();
+    protected  String networkName;
+    // mySelf is the current peer setting up the network
+    protected  SMSPeer mySelf;
+
+    //manager makes use of SMSHandler to send requests
+    private SMSHandler handler;
+
+    /**
+     * Set up a new network
+     *
+     * @param handler we set up a handler for sending requests
+     * @param networkName of the network being created
+     */
+    public void setup(SMSHandler handler, String networkName, SMSPeer mySelf){
+        this.handler = handler;
+        this.networkName = networkName;
+        this.mySelf = mySelf;
+        dict.addUser(mySelf);
+    }
+
+    /**
+     * Sends an invitation (JOIN_PROPOSAL request) to the specified peer
+     *
+     * @param peer who is asked to join the network
+     */
+    public void invite(SMSPeer peer){
+        SMSMessage invMsg = new SMSMessage(handler.getApplicationCode(), peer, JOIN_PROPOSAL + "_" + networkName);
+        joinSent.add(peer);
+        handler.sendMessage(invMsg);
+    }
+
+    /**
+     * Inform every user in the network that the current application is disconnecting from the network
+     * Sends a REMOVE_USER request, where user is mySelf
+     */
+
+    public void disconnect(){
+        spread(REMOVE_USER + "_" + mySelf.getAddress());
+        dict = null;
+        joinSent = null;
+        handler = null;
+    }
+
+    /**
+     * set a key-value resource for the local dictionary and spread this information to every user
+     *
+     * @param key resource key
+     * @param value resource value
+     */
+
+    public void setResource(SerializableObject key, SerializableObject value){
+        dict.setResource(key, value);
+        spread(ADD_RESOURCE + "_" + key.toString() + "_" + value.toString());
+    }
+
+    /**
+     * remove a key-value resource from the local dictionary and spread this information to every user
+     *
+     * @param key resource key
+     */
+
+    public void removeResource(SerializableObject key){
+        dict.removeResource(key);
+        spread(REMOVE_RESOURCE + "_" + key.toString());
+    }
+
+    /**
+     * Spreads this string according to your current index n in the dictionary. We send text to nodes with index
+     * 2n+1, 2n+2.
+     *
+     * @param text to be sent
+     */
+
+    private void spread(String text){
+        ArrayList<SMSPeer> users = dict.getAllUsers();
+        int first = 2*users.indexOf(mySelf) + 1;
+        for(int i = first; i < users.size() && i < first + 2; i++){
+            SMSMessage msg = new SMSMessage(handler.getApplicationCode(), users.get(i), text);
+            handler.sendMessage(msg);
+        }
+    }
+
```

Comment:
Sprending a message to 2n+1 and 2n+2, if the message starts from a user which is not number 0 in his users list, aren't you skipping the first elements, meaning they don't get notified?

Reply:
hai perfettamente ragione. Mi sono dimenticato del modulo

Reply:
Alright, I'd like to see this algorithm change to see how you can do it, other than that it's fine 👍